crates/subspace-archiving/src/archiver.rs

```diff
@@ -614,56 +631,82 @@ impl Archiver {
         Some(segment)
     }
 
-    /// Take segment as an input, apply necessary transformations and produce archived object mappings.
-    /// Must be called before `produce_archived_segment()`.
-    fn produce_object_mappings(&self, segment: &Segment) -> Vec<GlobalObject> {
+    /// Take the last block continuation in the next segment's items, apply necessary
+    /// transformations, and produce object mappings. Then remove the mappings from that block
+    /// continuation.
+    ///
+    /// Must be called on an incomplete segment.
+    fn produce_next_segment_mappings(&mut self) -> Vec<GlobalObject> {
+        let object_mapping = self.produce_object_mappings(self.buffer.iter());
+
+        // We've already produced these mappings, so we don't need to store them in the buffer.
+        if let Some(SegmentItem::BlockContinuation {
+            bytes: _,
+            object_mapping,
+        }) = self.buffer.back_mut()
+        {
+            object_mapping.objects_mut().clear();
```

Comment:
We wouldn't need to do additional matching here if we used `&mut [SegmentItems]` in `produce_object_mappings()` and drained objects there, which I'd argue is a good thing, we should need to produce mappings twice and releasing some RAM earlier shouldn't hurt.

Reply:
That isn’t possible as written, because we’d end up taking a mutable and immutable reference to the same Archiver on line 631.

But I can turn `produce_object_mappings()` into an associated method, because we’re only using the segment index from the Archiver directly.

Reply:
Yeah, even though we're not really doing anything problematic, it is not really possible to convince compile that is the case. Associated function it is then.